### ISSUE TITLE
Develop Actions In Item Struct

### ISSUE THREAD

**Issue by lindsey-hsi** [uchicago-cs/chiventure#215]
In this sprint, the newly implemented action struct within items must be further developed in close coordination with WDL and Action Management in order to account for conditions and effects + how they change the state of the game.

**Comment by a-matai** [comment]
After meeting WDL and Action Management, we have discussed the following theme:

action_t {
action_type_t
list of conditions (develop struct)
list of effects (develop struct)
_string_ text success 
_string_ text fail
_bool_ action allowed 
}

**Comment by patrickkoopmans** [comment]
Here are our preliminary ideas of how we will develop actions further. 
For record keeping sake, this feature is being developed on branch game-state/develop_actions. 

There will be a new hash table in the item_t struct called actions. We decided to add this hash table to distinguish between actions and attributes that have the same name. For example, a door might have an action “open” and an attribute “open”. 

Each action in the hash table will have an id that is its key. For example, a door will have the action “open”. Each key will correspond to a value that is of type game_action_t. 

Game_action_t is specified as follows: 
typedef struct action {
    char *action_name; 
    action_type_t *action_type; 
    action_condition_list_t; 
    action_effect_list_t; 
    char* success_str;
    char* fail_str;
} game_action_t;

The following explanation of this struct is based on a door item that currently has the attribute “open” with the value “no” so that you can see how the game state changes as we call the action to open the door:

•	action_name is the string name of the action, in this case we want to open the door so our action name is: “open” 

•	action_type is the action_type that is specified by action management in action_structs.h

•	action_condition_list is a linked list of game_action_condition structs. Each of these structs has two fields: the name of an attribute and the expected value of said attribute. In this example, the value of attribute “unlocked” will be expected to be “no” in order for the action “open” to proceed

•	action_effect_list is a linked list of game_effect_t structs. Each of these structs has three fields: the pointer to an item that is affected, the attribute of the item that is affected, and the new value of that attribute given the affect. In this case, the item “door” has an attribute called “open” which will be changed to “yes”

•	success_str is the string that prints should the action be successfully performed 

•	success_str is the string that prints should the action fail to be performed

•	Note: there was formerly a boolean member called “allowed”. This member has been deleted because it is better to check if all conditions are met each time an action is performed than to give it a permanent state of “allowed”. For example, we should check that the door is unlocked every time the player tries to open it rather than assume it has remained unlocked since the last time we opened it (the player might have pulled a sconce that locks the door behind them!). 

You might be thinking, what happens when the condition of an action relates to an attribute that is not of the item being acted upon. For example, in the previous example we only had to check if the door is unlocked, and we never had to check if the player was holding a key. So what would happen if the player needed to be in possession of a key to open an unlocked door? 

In this case, we would be dealing with a Type 3 action as specified by action management in action_structs.h. This is an action where we use one item (a key) to directly change the state of another item (a door). In this situation, we still do not add the condition that the player is holding a key to the action of opening the door. Rather, we store the effect of changing the “locked” attribute of “door” to “no” within the “key” action “use”. 

Thus, when action management uses their function do_item_item_action() they will first look for the action “use” in “key”. They will first check the conditions of performing this action. In this case, there are none. They will then iterate through the effects. As mentioned above, each effect has an item pointer, attribute name, and new attribute value. First, they will check that the item pointer is the same as the item pointer to “door”. This will ensure that they are using the key on the right door. Next, they will take the attribute name from the game_effect_t struct, and set it to the new attribute value (also found in the game_effect_t) using the functions provided in item.h. In this case, they will take the attribute “locked” and set it to “no” in the door. 

Thus, when they now call the action “open” on “door”, the door will already be unlocked and they can follow the steps that I outlined on the simple case of opening the door. 

It is most important to note that the door should not contain a condition of “key is in player inventory”. Rather the key will change the “locked” attribute of door, which will then satisfy its condition for opening. 

I hope this works for the WDL and AM teams, and I am more than happy to meet to discuss any concerns they might have.

**Comment by jessxia** [comment]
We removed the condition struct from path_t within the room module and modified or removed functions that referenced the condition_list_t within path to resolve errors in room.h and room.c.

We replaced the condition list in path_t with item_t *through to accommodate WDL's specifications for the item that a player must pass through to move between rooms.

**Comment by patrickkoopmans** [comment]
At this point, the game_effect_t and game_condition_t structs are written. We have modified/deleted tests that used the old model for actions, and noted down which old functions need to be re-implemented. 

We need to write the following functions:

//checks if a single condition of an action is met
bool check_condition(item_t *item, attribute_t* desired_attribute);

//checks if all conditions of an action are met
bool all_conditions_met(item_t* item, char* action_name)

/*add_allowed_action() adds a permitted action to an item
 * Parameters:
 *  a pointer to the item
 *  the action name
 *  the permitted action struct
 * Returns:
 *  SUCCESS if added correctly, FAILURE if failed to add
 */
int add_allowed_action(item_t* item, char *action_name, action_type_t *action_type, char* success_str, char* fail_str);

/* possible_action() (formerly allowed_action())checks if an action even exists for an item
 * Parameters:
 *  a pointer to the item
 *  the action name
 * Returns:
 *  SUCCESS if item contains action, FAILURE if it does not
 */
int possible_action(item_t* item, char* action_name);


We recommend action management write a perform_effect(game_action_effect* effect) function to use in do_item_action(), do_item_item_action()

**Comment by patrickkoopmans** [comment]
Functions written:
//checks a single condition
bool check_condition(item_t item, attribute_t desired_attribute);

//formerly add_allowed_action
int add_action(item_t item, char action_name, action_type_t action_type, char success_str, char fail_str);


Functions I have just thought of:
//will create a game_action_condition_t struct, initialize its values, and add it to an action
add_condition();

//will create a game_action_effect_t struct, initialize its values, and add it to an action
add_effect();


Functions left:
//checks if all conditions of an action are met
bool all_conditions_met(item_t* item, char* action_name)

//checks if an action even exists for an item
int possible_action(item_t item, char* action_name);

**Comment by patrickkoopmans** [comment]
Functions written:
//checks a single condition
bool check_condition(item_t item, attribute_t desired_attribute);

//formerly add_allowed_action
int add_action(item_t item, char action_name, action_type_t action_type, char success_str, char fail_str);

//checks if an action even exists for an item
bool possible_action(item_t item, char* action_name);


Functions left:
//checks if all conditions of an action are met
bool all_conditions_met(item_t* item, char* action_name)

//will create a game_action_condition_t struct, initialize its values, and add it to an action
add_condition();

//will create a game_action_effect_t struct, initialize its values, and add it to an action
add_effect();

**Comment by patrickkoopmans** [comment]
@borjasotomayor this is our "action development" issue that game-state referenced in class

**Comment by patrickkoopmans** [comment]
Notes post 06/03/19 after-class meeting with AM and WDL:
- Write "game_action_t get_action(item_t* item, char* action_name)" in item module -->DONE
- Delete action_type from item_t --> DONE
- Change action_list to action_hash in item_t --> DONE
- Write "bool path_cond_success(path_t* path)" function using game_action_condition struct within the path_t struct --> TODO
- Add back the functions that "disappeared" during branch from dev somehow (i.e. get_longdsec() from item) --> TODO

**Comment by MTUC08** [comment]
After starting implementation of AM functions with these functions, I have some additional requirements/requests on behalf of Action Management. 

REQUIRED: 
game_action_effect current looks like:
  game_action_effect {
       char * item_id;
       attribute_t *changed_attribute;
       struct game_action_effect *next;
}
but all of the set_attr functions require an item_t struct input, 
so game_action_effect should look like:
   game_action_effect {
       item_t *affected;
       attribute_t *changed_attribute;
       struct game_action_effect * next;
}

Additionally, within the attribute_tag enum, bool is BOOLE. Is this a typo? 


REQUESTS (NOT REQUIRED, BUT WOULD INCREASE CONVENIENCE)
Currently, to implement our functions, we are using a switch statement with the attribute_tag of the changed_attribute to determine which set_attr function to use. Adding a general set_attr function that takes in the item struct, attribute name, and a void field for the attribute value along with the attribute_tag and redirects to the correct function would be immensely useful in allowing abstraction.

**Comment by a-matai** [comment]
Hi Melinda, 

We are already in the process of dealing with the first request. 
Regd. the second, we will discuss and get back to you.

**Comment by a-matai** [comment]
Seeing as this bad boy is getting all filled up with various requests, we've decided to PR the action struct definitions etc into Master as soon as possible, and will then follow up with the rest of the functions so as not to have too large a PR drop on Monday. The goal is to have the basic action definitions and functions in develop_actions into dev by tomorrow and into master by sunday? (optimistic?).

**Updates to issue:**
- fixed functions according to new definitions (no more action_type_t in action structs, conditions now take in pointers to items instead)
- updated parameter input changes here and there
- created game_action.h header file that moved action/condition/effect functions while keeping definitions in item.h. Refactored to prevent "unwieldy" (Borja, 2019) header files.
 
**Unresolved for the newly defined PR**
- Do we still have char* success and failure in the action structs?
- Do we move all action functions from common-item to game_action.h?
- Is Action supposed to be a condition type? Are we removing this?
- What is the attributes_equal function supposed to do? Isn't this supposed to be what check_conditions is? Also does it return 1 on success and 0 on failure?
- Why am I still awake?
- What is your favourite ice cream flavour?

tldr:
this issue is being hijacked in the name of not tryna big PR anymore.
**All helper functions/modifying functions will be moved to a new issue. Please open up new issues for new requests!**

**Comment by MTUC08** [comment]
- You should keep char *success and failure because we are outputting those to CLI through AM functions

**Comment by MTUC08** [comment]
Also have you guys decided whether you will implement a general set_attr function?

**Comment by a-matai** [comment]
> Also have you guys decided whether you will implement a general set_attr function?

@MTUC08 I believe that initially we had done that, were asked to strong type them. 

Would it be possible for you to write one and call our different functions based on switch cases? We'll work on getting the other necessary functions up and running.

**Comment by MTUC08** [comment]
@a-matai Yeah that's what we currently have. I just thought that having a general set_attr might be better in terms of abstraction, but if you were told to strong type them that's totally understandable, no worries. To clarify, the other functions will all be in the game_action.h file?

**Comment by a-matai** [comment]
Merged dev into develop_actions
#nevercommit

**Comment by bansals10** [comment]
Hello from WDL! We were under the impression that this struct was being developed:
typedef struct action {
char action_name;
action_type_t action_type;
action_condition_list_t;
action_effect_list_t;
char* success_str;
char* fail_str;
} game_action_t;

1. There should be  char* success and failure in the action structs.
2. Moving all action functions from common-item to game_action.h would make sense.
3. Can you clarify what you mean by condition type? If you look at the updated WDL spec in master, a single action is a part of a list of actions which includes a list of conditions in association with each action.
4. If check_condition checks if the attribute is set to the correct value, then this is sufficient. From common.h:
#define SUCCESS (0)
#define FAILURE (1)
5. This is up to your implementation. As per our protocol, we creating a list which has a condition (obj) and a pointer to next.

**Comment by a-matai** [comment]
> @a-matai Yeah that's what we currently have. I just thought that having a general set_attr might be better in terms of abstraction, but if you were told to strong type them that's totally understandable, no worries. To clarify, the other functions will all be in the game_action.h file?

yep! @MTUC08

**Comment by a-matai** [comment]
> Hello from WDL! @a-matai
> 
> We were under the impression that this struct was being developed:
> typedef struct action {
> char action_name;
> action_type_t action_type;
> action_condition_list_t;
> action_effect_list_t;
> char* success_str;
> char* fail_str;
> } game_action_t;
> 
>     1. There should be  char* success and failure in the action structs.
> 
>     2. Moving all action functions from common-item to game_action.h would make sense.
> 
>     3. Can you clarify what you mean by condition type? If you look at the updated WDL spec in master, a single action is a part of a list of actions which includes a list of conditions in association with each action.
> 
>     4. If check_condition checks if the attribute is set to the correct value, then this is sufficient. From common.h:
>        #define SUCCESS (0)
>        #define FAILURE (1)
> 
>     5. This is up to your implementation. As per our protocol, we creating a list which has a condition (obj) and a pointer to next.

Yes that is the struct being developed! Sorry for being unclear, but some of the questions were for game-state! Namely 3, 4, 5!

**Comment by a-matai** [comment]
Updates to issue!
- Char* success/failure kept (the confusion around this was with AM/WDL and being unsure whether necessary)
- 
 
**Unresolved**
- Do we move all action functions from common-item to game_action.h? Specifically game_action_init/game_action_new? @liuyifan-eric 
- Why do conditions have a pointer to next inside them instead of later on creating a list which has a condition and a pointer to next? @liuyifan-eric

**Comment by a-matai** [comment]
Remaining to-do before PR into dev.
- Flesh out the init/new/add/possible functions

**Comment by jessxia** [comment]
In this issue (in item.h/item.c on the game-state/develop_actions branch), we defined the game_action_t struct, which contains:

- hash handle for hash table purposes
- string action name 
- linked list of conditions to be met in order to perform action
- linked list of effects caused by performance of action
- a success string to be printed upon success of action
- a failure string to be printed upon failure of action

The conditions and effects of an action are defined as game_action_condition_t and game_action_effect_t, and contain:

- a pointer to the item whose attribute we will be either looking at/changing
- a pointer to the attribute of the item that we want to check/change
- the expected or new value of the attribute
- a pointer to the next condition or effect in the list

Some initial functions that were written to accompany these structs are:

- check_condition, which checks if a condition is met in an item
- add_action, which adds an action to an item
- possible_action, which checks if an action is within the actions allowed for the item
- get_action, which returns an action from an item if it exists

Since our item module was getting bulky, we left the struct declarations in item but refactored the action functions to game_action.h and game_action.c. Another change that came with implementing actions within our item module was the removal of our initial condition struct within the room module and modification of functions that referenced the it. A field called through (consisting of an item pointer) was put into path to indicate the item that would be moved through (if applicable) between rooms.

All our questions in above comments have been resolved or moved to other issues. We are closing this issue even though the game_action module is underdeveloped in order to separate our work into smaller issues. Please see issue #315 for further functions for developing action conditions and #316 for action effects.

**Comment by mpcsh** [comment]
Issue Score: ✓+

Comments:
Great work!
